sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -1151,6 +1258,29 @@ def decision_function(self, X):
             decision = decision.ravel()
         return decision.T
 
+    def staged_decision_function(self, X):
+        """Compute decision function of ``X`` for each iteration.
+
+        This method allows monitoring (i.e. determine error on testing set)
+        after each stage.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The input samples.
+
+        Returns
+        -------
+        decision : generator of ndarray of shape (n_samples,) or \
+                (n_samples, n_trees_per_iteration)
+            The decision function of the input samples, which corresponds to
+            the raw values predicted from the trees of the ensemble . The
+            classes corresponds to that in the attribute :term:`classes_`.
+            Regression and binary classification are special cases with
+            ``k == 1``, otherwise ``k==n_classes``.
+        """
+        yield from self._staged_raw_predict(X)
```

Comment:
I haven't double checked but I think this should return the transpose?

Let's add a test for the returned shape of each method

Reply:
You are right. The _staged_raw_predict returns (n_trees_per_iteration, n_samples) while staged_decision_function should returns (n_samples, n_trees_per_iteration).

Reply:
To fix this, I will change to emulate decision_function for each staged_decision.
But the ```staged_predict_proba``` will use ```_staged_raw_predict ```directly.
That is, ```staged_decision_function``` will be separated from  ```staged_predict_proba``` and  ```staged_predict```

Reply:
why not just use


```py
for blah in self._staged_raw_predict(X):
	yield blah.T
```
?

Ideally we should keep all the `staged_XXX` as close as possible to their respective `XXX` counterpart.

Reply:
Yeah, I think I have kept all the staged_XXX as close as possible to the XXX counterpart.